From 05c451673aa176ebddb589851a28e509c87a2153 Mon Sep 17 00:00:00 2001
From: Murali Nalajala <mnalajal@codeaurora.org>
Date: Fri, 19 Oct 2012 18:47:27 +0530
Subject: [PATCH 17/75] msm: pm2: Configure MPA5_GDFS_CNT_VAL register for 8x25Q

Set the "DBGPWRUPEREQ_OVERRIDE" bits for C2 & C3. This allows
the cores to go into LPM when the debugger is connected.

Signed-off-by: Murali Nalajala <mnalajal@codeaurora.org>
(cherry picked from commit aa7310ba21fb6e7e64b95d6e955ea2c7104d6fe5)

Signed-off-by: Gopal Goberu <ggober@codeaurora.org>
(cherry picked from commit b17fd2d5df6c553c777977db3ad409726ec65609)

Change-Id: I5944095e93cbc386738b09e3227b11058c430d7d
Signed-off-by: Gopal Goberu <ggober@codeaurora.org>
---
 arch/arm/mach-msm/pm2.c |   35 ++++++++++++++++++++++++-----------
 1 files changed, 24 insertions(+), 11 deletions(-)

diff --git a/arch/arm/mach-msm/pm2.c b/arch/arm/mach-msm/pm2.c
index 7f57d53..cd7cfa6 100644
--- a/arch/arm/mach-msm/pm2.c
+++ b/arch/arm/mach-msm/pm2.c
@@ -980,17 +980,26 @@ static int msm_pm_power_collapse
 		/*
 		 * on system reset, default value of MPA5_GDFS_CNT_VAL
 		 * is = 0x0, later modem reprogram this value to
-		 * 0x00030004. Once APPS did a power collapse and
-		 * coming out of it expected value of this register
-		 * always be 0x00030004. Incase if APPS sees the value
-		 * as 0x00030002 consider this case as a modem early
-		 * exit.
+		 * 0x00030004/0x000F0004(8x25Q). Once APPS did
+		 * a power collapse and coming out of it expected value
+		 * of this register always be 0x00030004/0x000F0004(8x25Q).
+		 * Incase if APPS sees the value as 0x00030002/0x000F0002(8x25Q)
+		 * consider this case as a modem early exit.
 		 */
 		val = __raw_readl(MSM_CFG_CTL_BASE + 0x38);
-		if (val != 0x00030002)
-			power_collapsed = 1;
-		else
-			modem_early_exit = 1;
+
+		/* 8x25Q */
+		if (SOCINFO_VERSION_MAJOR(socinfo_get_version()) >= 3) {
+			if (val != 0x000F0002)
+				power_collapsed = 1;
+			else
+				modem_early_exit = 1;
+		} else {
+			if (val != 0x00030002)
+				power_collapsed = 1;
+			else
+				modem_early_exit = 1;
+		}
 	}
 
 #ifdef CONFIG_CACHE_L2X0
@@ -1681,12 +1690,16 @@ static int __init msm_pm_init(void)
 
 		/*
 		 * Configure the MPA5_GDFS_CNT_VAL register for
-		 * DBGPWRUPEREQ_OVERRIDE[17:16] = Override the
+		 * DBGPWRUPEREQ_OVERRIDE[19:16] = Override the
 		 * DBGNOPOWERDN for each cpu.
 		 * MPA5_GDFS_CNT_VAL[9:0] = Delay counter for
 		 * GDFS control.
 		 */
-		val = 0x00030002;
+		if (SOCINFO_VERSION_MAJOR(socinfo_get_version()) >= 3)
+			val = 0x000F0002;
+		else
+			val = 0x00030002;
+
 		__raw_writel(val, (MSM_CFG_CTL_BASE + 0x38));
 
 		l2x0_base_addr = MSM_L2CC_BASE;
-- 
1.7.0.4

